### PR TITLE
[Spec Constants] Improved handling of invalid spec. constants

### PR DESCRIPTION
**Pull request by RossBrunton**
Two main changes to how `Kernel/ProgramSetSpecializationConstants` are handled:
* They may now output either `INVALID_VALUE` or the new `INVALID_SPEC_ID` when the provided list is invalid.
* The OpenCL and level 0 adapters now respond to `UR_DEVICE_INFO_KERNEL_SET_SPECIALIZATION_CONSTANTS` with `false` rather than erroring out. This fixes some tests that were incorrectly not being skipped.
* `urKernelSetSpecializationConstants` now "implemented" (as a function that returns `UNSUPPORTED_FEATURE` for opencl and cuda.

### CODE REVIEW

File: test/conformance/program/program_adapter_cuda.match
Comment (aarongreig) [inline test/conformance/program/program_adapter_cuda.match:24]: I don't think any of the `ProgramSetSpecializationConstants` tests should be `{{OPT}}` for cuda and hip, neither support the entry point and we haven't gotten around to adding a device query to allow the tests to skip so they always fail.

**Review by aarongreig — APPROVED** [review]

**Comment by ldrumm** [comment]
@jchlanda is our spec constants expert. Jakub, please take over this review

File: source/adapters/cuda/kernel.cpp
Comment (jchlanda) [inline source/adapters/cuda/kernel.cpp:381]: Would it be worth it to do the same for HIP?

**Review by jchlanda — APPROVED** [review]

File: source/adapters/cuda/kernel.cpp
Comment (RossBrunton) [reply source/adapters/cuda/kernel.cpp:381]: HIP already does this.

File: include/ur_api.h
Comment (RossBrunton) [reply include/ur_api.h]: Reworded it to "A pSpecConstant entry contains a size that does not match that of the specialization constant in the module."

File: source/adapters/cuda/kernel.cpp
Comment (jchlanda) [reply source/adapters/cuda/kernel.cpp:381]: Thank you

File: include/ur_api.h
Comment (jchlanda) [reply include/ur_api.h]: Thank you.

**Review by PietroGhg — APPROVED** [review]
Native CPU lgtm, thank you

**Comment by kbenzie** [comment]
> @RossBrunton I think the HIP match files need updated on this.

Forgot you were off. I would make the changes to this PR myself but the checkbox to allow project maintainers to make changes appears to have been disabled.

**Comment by RossBrunton** [comment]
Can someone from @oneapi-src/unified-runtime-level-zero-write have a quick look at this? The changes for level_zero are changing an optional function from `DIE_NO_IMPLEMENTATION` to `  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;`

File: source/adapters/level_zero/device.cpp
Comment (nrspruit) [inline source/adapters/level_zero/device.cpp:865]: Shouldn't this be ur_bool_t{false}?

Comment (RossBrunton) [reply source/adapters/level_zero/device.cpp]: In the time between creating this MR and now, it looks as if this change has been added via 646cc9ce7 , so this MR no longer touches level zero. There's a lot of device infos that return `bool`s rather than `ur_bool_t` here, so it might be worth having a separate change that updates all of them at once.

**Comment by omarahmed1111** [comment]
@RossBrunton Please rebase this when possible, I removed the ready to merge label and please readd it when rebased.

**Comment by RossBrunton** [comment]
> @RossBrunton Please rebase this when possible, I removed the ready to merge label and please readd it when rebased.

Updated. Sorry for the delay.

**Pull request merged**
